### PR TITLE
mypy

### PR DESCRIPTION
**Pull request by oliver-sanders**
Now that we are starting to accumulate some type annotations in the codebase we ought to make sure that they are valid. No documentation is better than broken documentation.

This adds mypy to the CI and fixes outstanding issues to get it passing. We can add type annotations incrementally from here on.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Does not need tests (why?).
- [x] No change log entry required (why? e.g. invisible to users).
- [x] No documentation update required.
- [x] No dependency changes.

### CODE REVIEW

File: cylc/flow/c3mro.py [hunk 1]
```diff
@@ -150,7 +150,7 @@ def mro(self, C):
 
 
 if __name__ == "__main__":
-    parents = {}
+    parents: dict = {}
```
Comment (oliver-sanders) [inline cylc/flow/c3mro.py:153]: Have to add type annotations to the odd dict or two to keep mypy happy.

File: cylc/flow/data_messages_pb2.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# type: ignore
```
Comment (oliver-sanders) [inline cylc/flow/data_messages_pb2.py:1]: We can't configure per-file ignores at present so we have to stick a comment at the top of it.

File: cylc/flow/network/schema.py [hunk 3]
```diff
@@ -1455,10 +1455,11 @@ class Meta:
 
     class Arguments:
         workflows = List(WorkflowID, required=True)
+
         mode = BroadcastMode(
             # use the enum name as the default value
             # https://github.com/graphql-python/graphql-core-legacy/issues/166
-            default_value=BroadcastMode.Set.name,
+            default_value=BroadcastMode.Set.name,  # type: ignore
```
Comment (oliver-sanders) [inline cylc/flow/network/schema.py:1462]: mypy doesn't like Graphene enums.

File: cylc/flow/suite_files.py [hunk 4]
```diff
@@ -678,10 +678,10 @@ def remote_clean(reg, platform_names, timeout):
             if ret_code:
                 # Try again using the next platform for this install target:
                 this_platform = platforms.pop(0)
-                exc = TaskRemoteMgmtError(
+                excn = TaskRemoteMgmtError(
```
Comment (oliver-sanders) [inline cylc/flow/suite_files.py:681]: mypy doesn't like vars called `exc` unless they are in `except:` branches.

Comment (hjoliver) [reply cylc/flow/suite_files.py:681]: Huh, that's an interesting one

File: cylc/flow/network/graphql.py [hunk 5]
```diff
@@ -44,7 +45,7 @@
 
 STRIP_ARG = 'strip_null'
 NULL_VALUE = None
-EMPTY_VALUES = ([], {})
+EMPTY_VALUES: Tuple[list, dict] = ([], {})
```
Comment (MetRonnie) [inline cylc/flow/network/graphql.py:48]: Is it possible to be more specific here, e.g. `Tuple[List[int], Dict[str, str]]` or whatever it's meant to be?

File: cylc/flow/xtrigger_mgr.py [hunk 6]
```diff
@@ -126,16 +126,16 @@ def __init__(
         # For function arg templating.
         if not user:
             user = get_user()
-        self.farg_templ = {
+        self.farg_templ: dict = {
```
Comment (MetRonnie) [inline cylc/flow/xtrigger_mgr.py:129]:
```suggestion
        self.farg_templ: Dict[str, str] = {
```

File: cylc/flow/suite_files.py (hunk 4 above)
Comment (MetRonnie) [reply cylc/flow/suite_files.py:681]: I think it's because there's an earlier line `except Exception as exc:`. https://github.com/python/mypy/issues/5080

**Review by MetRonnie — COMMENTED** [review]
Instead of just `x: list = []` or `x: dict = {}`, I think it's more useful to specify the type of the entries, e.g. `x: List[str] = []` (needs `from typing import List`)

**Comment by oliver-sanders** [comment]
Agreed, however, the aim of this PR is to get `mypy` passing so that future type annotations are checked (so we don't add broken ones) rather than to upgrade the codebase now.

**Review by MetRonnie — APPROVED** [review]
Ok

**Review by hjoliver — APPROVED** [review]

**Pull request merged**